### PR TITLE
[WIP] Traditional flag parsing and eval flag

### PR DESCRIPTION
**Pull request by IanS5**
## Additions
This PR reworks flag parsing to be a bit more user-friendly, by implementing more traditional rules.

following capabilities are added:

- same argument short flags (i.e. `deno -rD` is legal,  and it is equivalent to `deno -r -D`, or `deno --reload --log-debug`)
- The special flag, `--`,  will immediately stop flag parsing
- short flag arguments may be in the same argv argument as the flag itself (i.e `deno -e'console.log("test")'` is the same as `deno -e 'console.log("test")'`)

This PR also adds the `-e` (`--eval`) flag, which can be passed a string to evaluate.

## Not Implemented

So far, no there is no error handling in the `set_flags` function, I've just left `unimplemented!()` where there should probably be some sort of error message or something in the future.

Evaluation is also not fully supported, each instance of `-e` is evaluated in a loop in the `main` function but what operations it can actually perform are _extremely_ limited, because it is run before `denoMain()`.

Deno still complains if it's missing an input file, even if one or more `-e` flags were provided

## Questions

I have a few questions about what error handling and the eval flag before they are implemented.

Should the error handling be done completely in the main set_flags function. So set_flags doesn't return and error, but instead prints an error message, or should it just return a `Result`.

Is eval practical at this point? I think it's a really helpful thing to have, and helps when just trying out deno, but is it worth it to change `denoMain()`, or use some other method to get it working properly?

### CODE REVIEW

File: src/fs.rs [hunk 1]
```diff
@@ -25,7 +25,10 @@ pub fn write_file_sync(path: &Path, content: &[u8]) -> std::io::Result<()> {
 
 pub fn mkdir(path: &Path) -> std::io::Result<()> {
   debug!("mkdir -p {}", path.display());
-  assert!(path.has_root(), "non-has_root not yet implemented");
+  assert!(
+    path.has_root(),
+    "non-has_root not yet implemented"
+  );
```
Comment (ry) [inline src/fs.rs]: Unrelated change?

File: src/handlers.rs [hunk 2]
```diff
@@ -33,15 +33,26 @@ pub extern "C" fn msg_from_js(d: *const DenoC, buf: deno_buf) {
       let msg = msg::CodeFetch::init_from_table(base.msg().unwrap());
       let module_specifier = msg.module_specifier().unwrap();
       let containing_file = msg.containing_file().unwrap();
-      handle_code_fetch(d, &mut builder, module_specifier, containing_file)
+      handle_code_fetch(
+        d,
+        &mut builder,
+        module_specifier,
+        containing_file,
+      )
```
Comment (ry) [inline src/handlers.rs]: Unrelated change?

File: src/handlers.rs [hunk 3]
```diff
@@ -33,15 +33,26 @@ pub extern "C" fn msg_from_js(d: *const DenoC, buf: deno_buf) {
...
       let source_code = msg.source_code().unwrap();
       let output_code = msg.output_code().unwrap();
-      handle_code_cache(d, &mut builder, filename, source_code, output_code)
+      handle_code_cache(
+        d,
+        &mut builder,
+        filename,
+        source_code,
+        output_code,
+      )
```
Comment (ry) [inline src/handlers.rs]: Unrelated change?

File: src/handlers.rs [hunk 4]
```diff
@@ -52,7 +63,13 @@ pub extern "C" fn msg_from_js(d: *const DenoC, buf: deno_buf) {
...
       // TODO base.msg_as_TimerStart();
       let msg = msg::TimerStart::init_from_table(base.msg().unwrap());
-      handle_timer_start(d, &mut builder, msg.id(), msg.interval(), msg.delay())
+      handle_timer_start(
+        d,
+        &mut builder,
+        msg.id(),
+        msg.interval(),
+        msg.delay(),
+      )
```
Comment (ry) [inline src/handlers.rs]: Unrelated change?

File: src/handlers.rs [hunk 5]
```diff
@@ -123,7 +140,11 @@ fn handle_start(
 ) -> HandlerResult {
   let deno = from_c(d);
 
-  let argv = deno.argv.iter().map(|s| s.as_str()).collect::<Vec<_>>();
+  let argv = deno
+    .argv
+    .iter()
+    .map(|s| s.as_str())
+    .collect::<Vec<_>>();
```
Comment (ry) [inline src/handlers.rs]: Unrelated change?

File: src/deno_dir.rs [hunk 6]
```diff
@@ -399,8 +408,17 @@ fn test_src_file_to_url() {
...
+    deno_dir
+      .deps
+      .join("hello/world.txt")
+      .to_str()
+      .unwrap(),
+  );
+  assert_eq!(
+    "http://hello/world.txt",
+    deno_dir.src_file_to_url(x)
+  );
```
Comment (ry) [inline src/deno_dir.rs]: Please don't include unrelated changes. (What version of rustfmt are you using? Mine doesn't do these.)

File: src/flags.rs [hunk 7]
```diff
@@ -106,6 +155,93 @@ fn test_set_flags_3() {
...
+      allow_net: false,
+      eval: svec!["console.error('ERROR: Oh no!')"],
+    }
+  );
+}
+
+#[test]
+fn test_set_flags_7() {
+  let (flags, rest) = set_flags(svec!["deno", "-h", "--", "-r", "-e"]);
+  assert!(rest == svec!["deno", "-r", "-e"]);
```
Comment (ry) [inline src/flags.rs]: Could you please leave out this test? I don't want to support "--" yet.

File: src/flags.rs [hunk 8]
```diff
@@ -44,20 +47,60 @@ pub fn set_flags(args: Vec<String>) -> (DenoFlags, Vec<String>) {
...
+        "--reload" => flags.reload = true,
+        "--eval" => {
+          match arg_iter.next() {
+            Some(e) => flags.eval.push(e.clone()),
+            None => unimplemented!(),
+          };
+        }
+        "--allow-write" => flags.allow_write = true,
+        "--allow-net" => flags.allow_net = true,
+        "--" => break,
```
Comment (ry) [inline src/flags.rs]: Could you please leave out the "--" case? I don't want to support "--" yet.

File: src/flags.rs [hunk 9]
```diff
@@ -19,19 +20,21 @@ pub struct DenoFlags {
...
---allow-net        Allow network access.
--v or --version    Print the version.
--r or --reload     Reload cached remote resources.
--D or --log-debug  Log debug output.
---help             Print this message.
---v8-options       Print V8 command line options."
+--allow-write          Allow file system write access.
+--allow-net            Allow network access.
+-v or --version        Print the version.
+-e or --eval [SCRIPT]  Evaluate a script from the command line, may be passed multiple times.
```
Comment (hayd) [inline src/flags.rs]: Do other languages support multiple eval scripts here? Neither python or node do.

Comment (IanS5) [reply src/flags.rs]: Perl does, it's not a big deal or anything, but personally, I like writing separate lines as `-e'...'` better than just separating them with a `;` in a single `eval` flag. Particularly in shell scripts, it can be kind of nice if you want to avoid a Here-Document. For example:

```bash
deno -e "console.log('$HELLO')" \
     -e "console.log('$WORLD')"
```

You could do something like this if you wanted to use env vars in a tiny script for some reason.

Comment (ry) [reply src/flags.rs]: Seems easy to support - I'm for it.

**Comment by ry** [comment]
Rebased and refactored in #675
